bson/src/main/org/bson/codecs/FlexibleCodecProvider.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bson.codecs;
+
+import org.bson.codecs.configuration.CodecProvider;
+import org.bson.codecs.configuration.CodecRegistry;
+
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * A {@code CodecProvider} for all codecs that can read in different values depending on
+ * the BsonRepresentation.
+ *
+ * @since 4.2
+ */
+public class FlexibleCodecProvider implements CodecProvider {
```

Comment:
This class is a little more complicated than it needs to be, but it copies the style of `BsonValueCodecProvider` because I'm assuming we're going to want to add support for additional BsonRepresentations in the future which will necessitate more FlexibleCodecs